backend/src/main/controller/PromotionController.ts

```diff
@@ -0,0 +1,188 @@
+import { NextFunction, Request, Response } from 'express';
+import { PromotionRepository } from '../repository/PromotionRepository';
+import { getCustomRepository, SelectQueryBuilder } from 'typeorm';
+import { Promotion } from '../entity/Promotion';
+import { UserRepository } from '../repository/UserRepository';
+import { Discount } from '../entity/Discount';
+import {
+  PromotionDTO,
+  PromotionValidation,
+} from '../validation/PromotionValidation';
+import {
+  PromotionQueryDTO,
+  PromotionQueryValidation,
+} from '../validation/PromotionQueryValidation';
+import { IdValidation } from '../validation/IdValidation';
+
+/* eslint-disable  @typescript-eslint/no-explicit-any */
+export class PromotionController {
+  /**
+   * Retrieves all promotions and their discounts
+   * * First we need to validate the query params and cast that into a PromotionQueryDTO
+   * * Then we apply the query options into the query builder depending on which properties are present
+   * * Now we execute the query builder and return its results
+   */
+  getAllPromotions = async (
+    request: Request,
+    response: Response,
+    next: NextFunction
+  ): Promise<any> => {
+    try {
+      let queryBuilder = getCustomRepository(PromotionRepository)
+        .createQueryBuilder('promotion')
+        .innerJoinAndSelect('promotion.discount', 'discount');
+
+      if (request.query) {
+        const promotionQuery: PromotionQueryDTO = await PromotionQueryValidation.schema.validateAsync(
+          request.query,
+          { abortEarly: false }
+        );
+        queryBuilder = this.applyQueryOptions(queryBuilder, promotionQuery);
+      }
+
+      const promotions = await queryBuilder.getMany();
+      return response.send(promotions);
+    } catch (e) {
+      return next(e);
+    }
+  };
+
+  /**
+   * Depending on which properties are defined inside promotionQuery, we add those properties into our query for the queryBuilder to execute.
+   */
+  private applyQueryOptions(
```

Comment:
This function attaches more to the query builder depending on any filters we apply for `GET promotions`.

Reply:
could be moved to `PromotionService` class, but I thought since its relevant to communicating w DB we can leave it here